File: js/demo/objects/menu.js
'use strict';
import { Drawable, GameObject, override } from '../../engine';

class Menu extends Drawable(GameObject) {
  opts = {};
  cur = 0;

  @override
  keydown(which) {
    switch(which) {
      case 'Enter':
        this.opts[Object.keys(this.opts)[this.cur]].call(this);
        break;
      case 'ArrowDown':
        this.cur = (this.cur + 1) % Object.keys(this.opts).length;
        break;
      case 'ArrowUp':
        this.cur = (this.cur - 1 + Object.keys(this.opts).length) % Object.keys(this.opts).length;
        break;
    }
  }

  @override
  drawGUI(draw) {
    Object.keys(this.opts).forEach((opt, i) => {
      draw
        .alpha(1)
        .color(i === this.cur ? 0xFF0000 : 0)
        .text(opt, [64, 128 + 32 * i]);
    });
  }
}

export default Menu;

File: c++/src/demo/objects/menu.cc
#include "menu.h"

namespace Demo {
    Menu::Menu(const std::vector<Option> & opts) : _opts { opts } {}

    void Menu::keydown(int key) {
        switch(key) {
            // TODO: don't use SDL scancodes... remap those or something for abstraction
            //       or maybe it's not worth it who knows...
        case SDL_SCANCODE_DOWN:
            _cur = (_cur + 1) % _opts.size();
            break;
        case SDL_SCANCODE_UP:
            _cur = (_cur + _opts.size() - 1) % _opts.size();
            break;
        case SDL_SCANCODE_RETURN:
            _opts.at(_cur).second();
            break;
        }
    }

    void Menu::draw(Game::Draw & draw) const {
        Game::Position where { 64, 128 };
        draw.font("hack");
        draw.alpha(1);
        for(unsigned int i = 0; i < _opts.size(); ++i) {
            auto & opt = _opts[i].first;
            where.y = 128 + 32 * i;
            if(i == _cur) {
                draw.color(0xff, 0, 0);
            } else {
                draw.color(0, 0, 0);
            }
            draw.text(opt, where);
        }
    }
}
